File: sorting/merge_sort/linked_list/c/linked_list.c
#include "linked_list.h"

typedef struct __node {
        struct object value;
        struct __node *next;
} node;

typedef struct __linked_list {
        obj_type type;
        size_t length;
        struct __node *head;
        struct __node *tail;
} linked_list;

static void increase_list_length(linked_list *list)
{
        list->length += 1;
}

static void decrease_list_length(linked_list *list)
{
        const size_t list_length = list->length;

        if (list_length > 0)
                list->length -= 1;
}

static node *get_node(linked_list *list, int index)
{
        if (!is_empty(list)) {
                const size_t list_length = list->length;
                const size_t list_end = list_length - 1;

                if (index > list_end || index < 0) {
                        fprintf(stderr, "***index [%d] out of bounds***\n", index);
                        destroy_linked_list(list);
                        exit(EXIT_FAILURE);
                }

                node *node = list->head;

                for (int i = 0; i < index; i++)
                        node = node->next;

                return node;
        }

        return NULL;
}

linked_list *new_linked_list(int type)
{
        linked_list *list = malloc(sizeof(linked_list));

        if (list == NULL)
                return NULL;

        obj_type list_type = get_object_type(type);

        *list = (linked_list) {
                .type = list_type,
                .length = 0,
                .head = NULL,
                .tail = NULL,
        };

        return list;
}

size_t get_length(linked_list *list)
{
        return list->length;
}

bool is_empty(linked_list *list)
{
        return list->head == NULL;
}

void insert_first(linked_list *list, object object)
{
        if (list != NULL) {
                node *new_node = malloc(sizeof(node));

                if (new_node != NULL) {
                        node *head = list->head;

                        *new_node = (node) {
                                .next = head,
                                .value = object,
                        };

                        if (list->tail == NULL)
                                list->tail = new_node;

                        list->head = new_node;

                        increase_list_length(list);
                }
        }
}

void insert_last(linked_list *list, object object)
{
        if (list != NULL) {
                node *new_node = malloc(sizeof(node));

                if (new_node != NULL) {
                        node *tail = list->tail;

                        if (!is_empty(list))
                                tail->next = new_node;

                        *new_node = (node) {
                                .next = NULL,
                                .value = object,
                        };

                        if (list->head == NULL)
                                list->head = new_node;

                        list->tail = new_node;

                        increase_list_length(list);
                }
        }
}

result remove_first(linked_list *list)
{
        if (!is_empty(list)) {
                node *head = list->head;

                if (get_length(list) == 1) {
                        list->head = NULL;
                        list->tail = NULL;
                } else {
                        list->head = head->next;
                }

                result obj = get_object(&head->value, list->type);

                decrease_list_length(list);
                free(head);

                return obj;
        }

        return (struct result) { .is_present = false };
}

result remove_last(linked_list *list)
{
        if (!is_empty(list)) {
                const size_t list_length = list->length;
                const size_t list_end = list_length - 1;

                node *tail = NULL;

                if (get_length(list) == 1) {
                        tail = list->head;

                        list->head = NULL;
                        list->tail = NULL;
                } else {
                        node *previous = get_node(list, list_end - 1);

                        tail = previous->next;
                        previous->next = NULL;
                }

                result obj = get_object(&tail->value, list->type);

                decrease_list_length(list);
                free(tail);

                return obj;
        }

        return (struct result) { .is_present = false };
}

void insert_obj_at(linked_list *list, object object, int index)
{
        if (list != NULL) {
                const size_t list_length = list->length;
                const size_t list_end = list_length - 1;

                if (index > list_length || index < 0) {
                        fprintf(stderr, "***index [%d] out of bounds***\n", index);
                        destroy_linked_list(list);
                        exit(EXIT_FAILURE);
                }

                if (index == 0) {
                        insert_first(list, object);
                } else if (index == list_end) {
                        insert_last(list, object);
                } else {
                        node *previous = get_node(list, index - 1);
                        node *current = previous->next;

                        node *new_node = malloc(sizeof(node));

                        if (new_node != NULL) {
                                *new_node = (node) {
                                        .next = current,
                                        .value = object,
                                };

                                previous->next = new_node;

                                if (list->tail == NULL)
                                        list->tail = new_node;

                                increase_list_length(list);
                        }
                }
        }
}

result get_obj_at(linked_list *list, int index)
{
        if (!is_empty(list)) {
                const size_t list_length = list->length;
                const size_t list_end = list_length - 1;
                const int list_type = list->type;

                if (index > list_end || index < 0) {
                        fprintf(stderr, "***index [%d] out of bounds***\n", index);
                        destroy_linked_list(list);
                        exit(EXIT_FAILURE);
                }

                node *node = get_node(list, index);

                return get_object(&node->value, list_type);
        }

        return (struct result) { .is_present = false };
}

result remove_obj_at(linked_list *list, int index)
{
        if (!is_empty(list)) {
                const size_t list_length = list->length;
                const size_t list_end = list_length - 1;

                if (index > list_end || index < 0) {
                        fprintf(stderr, "***index [%d] out of bounds***\n", index);
                        destroy_linked_list(list);
                        exit(EXIT_FAILURE);
                }

                if (index == 0) {
                        remove_first(list);
                } else if (index == list_end) {
                        remove_last(list);
                } else {
                        node *previous = get_node(list, index - 1);
                        node *current = previous->next;

                        previous->next = current->next;

                        result obj = get_object(&current->value, list->type);

                        free(current);
                        decrease_list_length(list);

                        return obj;
                }
        }

        return (struct result) { .is_present = false };
}

void show_list(linked_list *list, to_string_fn to_string, bool reverse)
{
        if (!is_empty(list)) {
                const size_t list_length = list->length;

                int start = reverse ? list_length : 0;
                int end = reverse ? 0 : list_length;

                if (reverse) {
                        for (int i = start - 1; i >= end; i--)
                                to_string(get_obj_at(list, i).object);
                } else {
                        for (int i = start; i < end; i++)
                                to_string(get_obj_at(list, i).object);
                }
        }
}

static node *merge(node *first_half, node *second_half, compare_obj_fn comparator)
{
        node *result = NULL;

        if (first_half == NULL)
                return second_half;
        else if (second_half == NULL)
                return first_half;

        if (comparator(first_half->value, second_half->value) <= 0) {
                result = first_half;
                result->next = merge(first_half->next, second_half, comparator);
        } else {
                result = second_half;
                result->next = merge(first_half, second_half->next, comparator);
        }

        return result;
}

static void split_list(node *src, node **first_half, node **second_half)
{
        node *aux_1;
        node *aux_2;

        aux_2 = src;
        aux_1 = src->next;

        while (aux_1 != NULL) {
                aux_1 = aux_1->next;
                if (aux_1 != NULL) {
                        aux_2 = aux_2->next;
                        aux_1 = aux_1->next;
                }
        }

        *first_half = src;
        *second_half = aux_2->next;
        aux_2->next = NULL;
}

static void merge_sort(node **start, compare_obj_fn comparator)
{
        node *head = *start;

        if (head == NULL || head->next == NULL)
                return;

        node *first_half = NULL;
        node *second_half = NULL;

        split_list(head, &first_half, &second_half);

        merge_sort(&first_half, comparator);
        merge_sort(&second_half, comparator);

        *start = merge(first_half, second_half, comparator);
}

void m_sort_list(linked_list *list, compare_obj_fn comparator)
{
        if (is_empty(list))
                return;

        merge_sort(&list->head, comparator);
}

void destroy_linked_list(linked_list *list)
{
        if (list != NULL) {
                while (!is_empty(list))
                        remove_first(list);

                list->type = -1;
                list->length = -1;
                list->head = NULL;
                list->tail = NULL;

                free(list);
                list = NULL;
        }
}

File: linked_list/go/linkedlist/linkedlist.go
package linkedlist

import (
	"errors"
	"fmt"
	"sync"
)

var (
	errIndexOutOfBounds = errors.New("index out of bounds")
)

type node struct {
	Data interface{}
	Next *node
}

type linkedList struct {
	length int
	head   *node
	tail   *node
	lock   sync.Mutex
}

func newNode(data interface{}, next *node) *node {
	return &node{
		Data: data,
		Next: next,
	}
}

func NewLinkedList() *linkedList {
	return &linkedList{}
}

func (l *linkedList) increaselength() {
	l.length += 1
}

func (l *linkedList) decreaselength() {
	if l.length > 0 {
		l.length -= 1
	}
}

func (l *linkedList) findNode(index int) (*node, error) {
	if !l.IsEmpty() {
		if index < 0 && index > l.length {
			return nil, errIndexOutOfBounds
		}

		node := l.head
		for i := 0; i < index; i++ {
			node = node.Next
		}

		return node, nil
	}

	return nil, nil
}

func (l *linkedList) Getlength() int {
	return l.length
}

func (l *linkedList) IsEmpty() bool {
	return l.head == nil
}

func (l *linkedList) InsertFirst(data interface{}) {
	l.lock.Lock()
	defer l.lock.Unlock()

	newNode := newNode(data, l.head)

	l.head = newNode

	if l.tail == nil {
		l.tail = newNode
	}

	l.increaselength()
}

func (l *linkedList) InsertLast(data interface{}) {
	l.lock.Lock()
	defer l.lock.Unlock()

	newNode := newNode(data, nil)

	if !l.IsEmpty() {
		l.tail.Next = newNode
	}

	l.tail = newNode

	if l.head == nil {
		l.head = newNode
	}

	l.increaselength()
}

func (l *linkedList) InsertAt(data interface{}, index int) error {

	if index < 0 || index > l.length {
		return errIndexOutOfBounds
	}

	switch index {
	case 0:
		l.InsertFirst(data)
	case l.length:
		l.InsertLast(data)
	default:
		l.lock.Lock()
		defer l.lock.Unlock()

		prev, _ := l.findNode(index - 1)
		newNode := newNode(data, prev.Next)
		prev.Next = newNode

		l.increaselength()
	}

	return nil
}

func (l *linkedList) RemoveFirst() interface{} {
	l.lock.Lock()
	defer l.lock.Unlock()

	if !l.IsEmpty() {
		head := l.head

		if l.length == 1 {
			l.head = nil
			l.tail = nil
		} else {
			l.head = head.Next
		}

		l.decreaselength()

		return head.Data
	}

	return nil
}

func (l *linkedList) RemoveLast() interface{} {
	l.lock.Lock()
	defer l.lock.Unlock()

	if !l.IsEmpty() {
		tail := l.tail

		if l.length == 1 {
			l.head = nil
			l.tail = nil
		} else {
			listEnd := l.length - 1
			prev, _ := l.findNode(listEnd - 1)
			prev.Next = nil
			l.tail = prev
		}

		l.decreaselength()

		return tail.Data
	}

	return nil
}

func (l *linkedList) RemoveAt(index int) (interface{}, error) {
	if index < 0 || index >= l.length {
		return nil, errIndexOutOfBounds
	}

	if !l.IsEmpty() {
		switch index {
		case 0:
			data := l.RemoveFirst()
			return data, nil
		case l.length - 1:
			data := l.RemoveLast()
			return data, nil
		default:
			l.lock.Lock()
			defer l.lock.Unlock()

			prev, _ := l.findNode(index - 1)
			current, _ := l.findNode(index)
			prev.Next = current.Next
			l.decreaselength()

			return current.Data, nil
		}
	}

	return nil, nil
}

func (l *linkedList) Get(index int) (interface{}, error) {
	if index < 0 || index >= l.length {
		return nil, errIndexOutOfBounds
	}

	if !l.IsEmpty() {
		node, _ := l.findNode(index)
		return node.Data, nil
	}

	return nil, nil
}

func (l *linkedList) Show() {
	if !l.IsEmpty() {
		for node := l.head; node != nil; node = node.Next {
			fmt.Println(node.Data)
		}
	}
}

func (l *linkedList) Clear() {
	l.lock.Lock()
	defer l.lock.Unlock()

	if !l.IsEmpty() {
		l.head = nil
		l.tail = nil
		l.length = 0
	}
}

func (l *linkedList) Reverse() {
	l.lock.Lock()
	defer l.lock.Unlock()

	if !l.IsEmpty() {
		l.tail = l.head

		var prev, next *node
		current := l.head
		for current != nil {
			next = current.Next
			current.Next = prev
			prev = current
			current = next
		}

		l.head = prev
	}
}
